sources/packages/backend/libs/services/src/notifications/notification/notification.service.ts

```diff
@@ -70,57 +80,130 @@ export class NotificationService extends RecordDataModelService<Notification> {
         id: notificationId,
       },
-      { dateSent: new Date() },
+      { dateSent: new Date(), permanentFailureError },
     );
   }
 
   /**
-   * Invokes the sendEmailNotification method of the GC Notification service.
-   * @param notificationId notification id used to retrieve the payload that will be passed
-   * as a parameter to the sendMailNotification method.
-   * @param entityManager optional repository that can be provided, for instance,
-   * to execute the command as part of an existing transaction. If not provided
-   * the local repository will be used instead.
-   * @returns GC Notify API call response.
+   * Process all the unsent notifications with a polling limit.
+   * Processing continues recursively until all the records are processed.
+   * Call GCNotify to send email notification.
+   * @param pollingLimit
+   * @param notificationsProcessed optional param used to
+   * increment count on recursion.
+   * @param notificationsSuccessfullyProcessed optional param used to
+   * increment count on recursion.
+   * @returns processing summary.
    */
-  async sendEmailNotification(
-    notificationId: number,
-    entityManager?: EntityManager,
-  ): Promise<GCNotifyResult> {
-    const notificationRepo =
-      entityManager?.getRepository(Notification) ?? this.repo;
-    const notification = await notificationRepo.findOne({
-      select: {
-        id: true,
-        messagePayload: true,
-      },
-      where: {
-        id: notificationId,
-      },
-    });
-    // Call GC Notify send email method.
-    const gcNotifyResult = await this.gcNotifyService.sendEmailNotification(
-      notification.messagePayload,
-    );
+  async processUnsentNotifications(
+    pollingLimit: number,
+    notificationsProcessed = 0,
+    notificationsSuccessfullyProcessed = 0,
+  ): Promise<NotificationProcessingStatus> {
+    const notificationsToProcess = await this.repo
+      .createQueryBuilder("notification")
+      .select(["notification.id", "notification.messagePayload"])
+      .where("notification.dateSent IS NULL")
+      .limit(pollingLimit)
+      .getMany();
+
+    if (notificationsToProcess.length) {
+      this.logger.log(
+        `Processing ${notificationsToProcess.length} notifications`,
+      );
+
+      // Used to limit the number of asynchronous operations
+      // that will start at the same time.
+      const maxPromisesAllowed = os.cpus().length;
+
+      // Hold all the promises that must be processed.
+      const promises: Promise<boolean>[] = [];
+
+      for (const notification of notificationsToProcess) {
+        promises.push(this.sendEmailNotification(notification));
 
-    // Update date sent column in notification table after sending email notification successfully.
-    await this.updateNotification(notification.id, entityManager);
+        if (promises.length >= maxPromisesAllowed) {
+          // Waits for all be executed.
+          const processingResult = await Promise.all(promises);
```

Comment:
We talked about avoiding using the `Promise.all` and using the `Promise.allSettled()`, does anything changed from the last conversation?

Reply:
Yes, now that we are not throwing error at all and failing safe all the times, went for Promise.all instead of all settled.

Reply:
True, I guess that we talked about it and I forgot.
Just to keep in mind, yes, there is a try/catch inside the `sendEmailNotification` where another error can happen and force the process to be aborted since the catch also do processing that it is not in a `try/catch`.
My point is, `Promise.allSettled()` would be the best choice for this type of process.